### PR TITLE
Fr1  -- RequireFinalReturn

### PR DESCRIPTION
**Pull request by ziuq**
Silencing the require final return criticism.  I was a believer in this critique before, now I believe more deeply.

Added a lot of  returns.   Removed one or two function prototypes and a few trailing spaces.

In 9e69fc32, removed  duplicated bulk from the branches of a condition.

This has passed local tests and TravisCI.

Rob

### CODE REVIEW

File: lib/LedgerSMB/File.pm [hunk 1]
```diff
@@ -306,7 +307,7 @@ sub DEMOLISH {
        unlink $self->{file_path} . '/' . $file;
    }
    closedir (TMP);
-   rmdir $self->{file_path};
+   return rmdir $self->{file_path};
```
Comment (ehuelsmann) [inline lib/LedgerSMB/File.pm]: this probably didn't mean to return the rmdir value, but 'undef' instead.

File: lib/LedgerSMB/Report/Dates.pm [hunk 2]
```diff
@@ -216,7 +216,7 @@ sub _get_to_date {
     if ($self->interval eq 'month'){
        $date->add(months => 1);
     } elsif ($self->interval eq 'quarter'){
-       $date->add(months => 3);
+       $date->add(months => 3);         ## no critic ( ProhibitMagicNumbers)
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Report/Dates.pm]: Hmm. Where did the ProhibitMagicNumbers come from?

File: lib/LedgerSMB/Scripts/setup.pm [hunk 3]
```diff
@@ -248,15 +248,15 @@ sub login {
...
+    return $template->render_to_psgi($request);
 }
 
 =item sanity_checks
 Checks for common setup issues and errors if admin tasks cannot be completed/
 
 =cut
 
-sub sanity_checks {
+sub sanity_checks { ## no critic ( RequireFinalReturn )
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: I can see that we want to return 'undef' (or a simple  final `return;`) here: this seems to be more of a `procedure` (Pascal terms: not meant to have a return value) than a `function` (explicit return value)

File: lib/LedgerSMB/Template.pm [hunk 4]
```diff
@@ -619,7 +620,7 @@ sub _http_output {
     print $data;
     # change global resource back asap
     binmode (STDOUT, ':utf8');
-    $logger->trace("end print to STDOUT");
+    return $logger->trace("end print to STDOUT");
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Template.pm]: There's an empty `return;` meant to be closing this sub.

File: lib/LedgerSMB/Template.pm [hunk 5]
```diff
@@ -693,7 +696,7 @@ sub _email_output {
             @attachment,
         );
     }
-    $mail->send;
+    return $mail->send;
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Template.pm]: There's an empty `return;` meant to be here, I think.

File: lib/LedgerSMB/Template.pm [hunk 6]
```diff
@@ -718,7 +721,7 @@ sub _lpr_output {
     }
 
     close $pipe;
-    close $file;
+    return close $file;
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Template.pm]: Same empty `return;` here too, I think.

File: lib/LedgerSMB/Template/ODS.pm [hunk 7]
```diff
@@ -147,11 +147,12 @@ sub _worksheet_handler {
     } else {
         $sheet = $ods->appendTable($_->{att}->{name}, $rows, $columns);
     }
+    return;
 }
 
 sub _row_handler {
     $rowcount++;
-    $currcol = 0;
+    return $currcol = 0;
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Template/ODS.pm]: This callback and the ones below are probably expected to return undef: `return;`

File: lib/LedgerSMB/Template/XLS.pm [hunk 8]
```diff
@@ -67,25 +67,25 @@ my $currcol;
 sub _worksheet_handler {
     $rowcount = -1;
     $currcol = 0;
-    $_->set_att(type => 'worksheet');
+    return $_->set_att(type => 'worksheet');
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Template/XLS.pm]: These handlers probably should be returning undef (`return;`)

File: lib/LedgerSMB/Template/XLSX.pm [hunk 9]
```diff
@@ -67,25 +67,25 @@ my $currcol;
 sub _worksheet_handler {
     $rowcount = -1;
     $currcol = 0;
-    $_->set_att(type => 'worksheet');
+    return $_->set_att(type => 'worksheet');
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Template/XLSX.pm]: These handlers probably should be returning undef (`return;`)

File: lib/LedgerSMB/File.pm (hunk 1 above)
Comment (ziuq) [inline lib/LedgerSMB/File.pm]: In the face of 25% test coverage, usually no doc of the return values, and no response on lsmb.dev, I mostly preserved the actual logic of the code, rather than go with my guess of the author's intention.  20% bad guesses * 20% bad luck on those * 25% covered leaves a couple of unnoticed bugs alive in code.  

I varied from conservative to very conservative on risking logic changes.    I did change loop-condition-returns and where an ending conditional gave very disparate final values.

So I agree with all your 'probably meant "return;"', but used a more conservative standard.
I have no problem with redoing any or all of these on your say so.

Comment (ehuelsmann) [inline lib/LedgerSMB/File.pm]: Please don't take my review as something negative. It's definitely not something ad-hominem: the feedback I'm providing merely serves to improve the change we'll be merging to the master branch. That might include changes which - while staring us in the face due to this change - should be fixed one time or another. It's more practical to do that right now with the PR pending, because now they can't be forgotten.

In short: yes, I'd like you to make the changes.

File: lib/LedgerSMB/Report/Dates.pm (hunk 2 above)
Comment (ziuq) [inline lib/LedgerSMB/Report/Dates.pm]: Sorry about that.  I meant to mention it in the comment above.  At one point, I wanted something to push to Github this was it. 

Do you think that in context this is not very magical and so no-critic works?  MONTHS_PER_QUARTER is more correct.  Which is preferred,  Readonly or constant or
something else?

File: lib/LedgerSMB/File.pm (hunk 1 above)
Comment (ziuq) [inline lib/LedgerSMB/File.pm]: No offense taken.  I was asking on lsmb dev for guidance on this, I will certainly take it now.

All your mentioned changes are up for review, plus a few cousins that show a strong family resemblance..

**Comment by ehuelsmann** [comment]
Thanks for the contribution! Merging.
